File: 05_analizar_emocion_odio_ironia.R
## 05_analizar_emocion_odio_ironia.R
##
## Versión R (vía reticulate) de 05_analizar_emocion_odio_ironia.py.
## Llama exactamente a los mismos analizadores de pysentimiento (Python)
## desde R, sin reescribir el modelo -reticulate solo traduce la llamada.
##
## Requisitos (una sola vez):
##   install.packages("reticulate")
##   reticulate::conda_create("pysentimiento_env", python_version = "3.10")
##   reticulate::conda_install("pysentimiento_env",
##                              packages = c("pysentimiento", "torch", "pandas", "tqdm"),
##                              pip = TRUE)
##
## Input : dataset_unido.csv (generado por 01_preparar_datos.py)
## Output: analisis_emocion_odio_ironia.csv

library(reticulate)
library(dplyr)
library(readr)
library(purrr)
library(tibble)

use_condaenv("pysentimiento_env", required = TRUE)
ps <- import("pysentimiento")

# ------------------------------------------------------------------
# Configuración
# ------------------------------------------------------------------
INPUT_PATH        <- Sys.getenv("INPUT_PATH", "dataset_unido.csv")
OUTPUT_PATH       <- Sys.getenv("OUTPUT_PATH", "analisis_emocion_odio_ironia.csv")
TASKS             <- c("emotion", "hate_speech", "irony")
LANG              <- "es"
BATCH_SIZE        <- as.integer(Sys.getenv("BATCH_SIZE", 32))
CHECKPOINT_EVERY  <- as.integer(Sys.getenv("CHECKPOINT_EVERY", 200))

sanitize <- function(nombre) {
  nombre <- trimws(tolower(nombre))
  gsub(" ", "_", nombre, fixed = TRUE)
}

clasificar_lote <- function(analyzers, textos) {
  # Une, en un solo dataframe (una fila por texto), las probabilidades y la
  # etiqueta derivada de las 3 tareas. Mismo criterio que la versión Python:
  # single-label -> "<tarea>_etiqueta" (un valor); multi-label ->
  # "<tarea>_etiquetas" (lista separada por coma, o "ninguna").
  filas <- vector("list", length(textos))
  for (i in seq_along(filas)) filas[[i]] <- list()

  for (task in names(analyzers)) {
    analyzer <- analyzers[[task]]
    # as.list() es clave: un vector de largo 1 se convertiría en un str de
    # Python (no una lista), y predict() devolvería un solo objeto en vez
    # de una lista -rompiendo la indexación de abajo en el último lote
    # parcial. as.list() fuerza siempre una lista de Python.
    salidas <- analyzer$predict(as.list(textos))

    for (i in seq_along(textos)) {
      salida <- salidas[[i]]
      # reticulate ya convierte automáticamente $probas (dict -> lista con
      # nombres), $is_multilabel (bool -> logical) y $output (str o list ->
      # character) al acceder con "$", así que no hace falta llamar
      # py_to_r() explícitamente sobre ellos.
      probas <- salida$probas
      for (label in names(probas)) {
        col <- paste0("prob_", task, "_", sanitize(label))
        filas[[i]][[col]] <- probas[[label]]
      }
      es_multilabel <- salida$is_multilabel
      if (es_multilabel) {
        activas <- salida$output
        col <- paste0(task, "_etiquetas")
        filas[[i]][[col]] <- if (length(activas) == 0) "ninguna" else paste(activas, collapse = ", ")
      } else {
        col <- paste0(task, "_etiqueta")
        filas[[i]][[col]] <- salida$output
      }
    }
  }
  filas
}

filas_a_tibble <- function(filas_resultado) {
  bind_rows(lapply(filas_resultado, as_tibble))
}

main <- function() {
  df <- read_csv(INPUT_PATH, show_col_types = FALSE)
  n_total <- nrow(df)
  cat(sprintf("Publicaciones a clasificar: %d\n", n_total))
  cat(sprintf("Tareas: %s (lang=%s)\n", paste(TASKS, collapse = ", "), LANG))

  textos <- as.character(df$descripcion)

  # --- Retomar desde checkpoint si existe ---
  checkpoint_path <- paste0(OUTPUT_PATH, ".checkpoint.csv")
  filas_resultado <- list()
  inicio <- 0

  if (file.exists(checkpoint_path)) {
    parcial_previo <- read_csv(checkpoint_path, show_col_types = FALSE)
    columnas_nuevas <- setdiff(names(parcial_previo), names(df))
    if (nrow(parcial_previo) <= n_total && length(columnas_nuevas) > 0) {
      inicio <- nrow(parcial_previo)
      filas_resultado <- pmap(parcial_previo[columnas_nuevas], function(...) list(...))
      cat(sprintf("Checkpoint encontrado: retomando desde la publicación %d/%d.\n", inicio, n_total))
    } else {
      cat("Aviso: el checkpoint no coincide con el dataset actual; se ignora y se parte de cero.\n")
    }
  }

  if (inicio >= n_total) {
    cat("El checkpoint ya tiene todas las publicaciones clasificadas; solo falta consolidar el resultado.\n")
  }

  t0 <- Sys.time()
  analyzers <- setNames(
    lapply(TASKS, function(task) ps$create_analyzer(task = task, lang = LANG, batch_size = BATCH_SIZE)),
    TASKS
  )
  cat(sprintf("Analizadores cargados en %.1fs\n", as.numeric(Sys.time() - t0, units = "secs")))

  batches_por_checkpoint <- max(1L, CHECKPOINT_EVERY %/% BATCH_SIZE)
  contador_batches <- 0
  # Nota: R es 1-indexado (a diferencia de Python), por eso el rango de
  # inicio y el largo del último lote parcial se calculan distinto.
  starts <- seq(inicio + 1, n_total, by = BATCH_SIZE)

  t0 <- Sys.time()
  pb <- txtProgressBar(min = 0, max = max(length(starts), 1), style = 3)
  for (k in seq_along(starts)) {
    start <- starts[k]
    fin <- min(start + BATCH_SIZE - 1, n_total)
    lote <- textos[start:fin]
    filas_resultado <- c(filas_resultado, clasificar_lote(analyzers, lote))
    contador_batches <- contador_batches + 1
    setTxtProgressBar(pb, k)

    if (contador_batches %% batches_por_checkpoint == 0) {
      parcial <- df[seq_len(length(filas_resultado)), ]
      parcial <- bind_cols(parcial, filas_a_tibble(filas_resultado))
      write_excel_csv(parcial, checkpoint_path)  # utf-8 con BOM, como utf-8-sig en pandas
    }
  }
  close(pb)
  cat(sprintf("\nClasificación completa en %.1f min\n", as.numeric(Sys.time() - t0, units = "mins")))

  df <- bind_cols(df, filas_a_tibble(filas_resultado))
  write_excel_csv(df, OUTPUT_PATH)
  cat(sprintf("Guardado: %s\n", OUTPUT_PATH))

  if (file.exists(checkpoint_path)) file.remove(checkpoint_path)

  cat("\nDistribución de emotion_etiqueta:\n")
  print(table(df$emotion_etiqueta))

  cat("\nDistribución de irony_etiqueta:\n")
  print(table(df$irony_etiqueta))

  cat("\nFrecuencia de cada categoría de hate_speech (multi-etiqueta, no excluyentes):\n")
  for (label in c("hateful", "targeted", "aggressive")) {
    col <- paste0("prob_hate_speech_", label)
    if (col %in% names(df)) {
      n <- sum(df[[col]] > 0.5)
      cat(sprintf("  %s: %d (%.1f%%)\n", label, n, 100 * n / nrow(df)))
    }
  }
}

if (sys.nframe() == 0) {
  main()
}

File: 05_analizar_emocion_odio_ironia.py
"""
05_analizar_emocion_odio_ironia.py

Clasificación de las descripciones de Instagram con los analizadores en
español de la librería pysentimiento (modelos RoBERTuito), para las tareas:

  - emotion      (pysentimiento/robertuito-emotion-analysis)
        7 clases, single-label (softmax, excluyentes entre sí):
        others, joy, sadness, anger, surprise, disgust, fear.

  - hate_speech  (pysentimiento/robertuito-hate-speech)
        3 categorías NO excluyentes, multi-label (sigmoide independiente
        por categoría -mismo espíritu que las 12 hipótesis NLI de
        02_clasificar_hipotesis.py-): hateful, targeted, aggressive.
        Basado en el esquema HatEval: "hateful" = es discurso de odio,
        "targeted" = ataca a un individuo específico (no a un grupo en
        general), "aggressive" = usa lenguaje agresivo/ofensivo.
        pysentimiento decide internamente qué categorías quedan
        "activas" con el mismo criterio simple de umbral 0.5 aplicado a
        cada probabilidad por separado (ver AnalyzerOutput en la
        librería) -no hay nada adicional ahí, igual que con UMBRAL en
        02_clasificar_hipotesis.py.

  - irony        (pysentimiento/robertuito-irony)
        2 clases, single-label (softmax): "ironic", "not ironic".

No incluye la tarea general "sentiment" (POS/NEG/NEU) de pysentimiento
porque no fue pedida. Si se quiere agregar después, basta con sumar
"sentiment" a la lista TASKS de abajo -el resto del script funciona
igual sin más cambios, ya que las columnas de salida se arman
dinámicamente a partir de lo que cada analizador devuelve.

------------------------------------------------------------------------
NOTA METODOLÓGICA
------------------------------------------------------------------------
Los modelos RoBERTuito de pysentimiento fueron preentrenados y afinados
sobre tuits en español (no sobre descripciones de Instagram), pero su
registro (texto corto, hashtags, emojis, menciones) es razonablemente
cercano al de las descripciones de este corpus. La función interna
preprocess_tweet() ya normaliza URLs, menciones y hashtags antes de
clasificar, así que no hace falta preprocesar el texto a mano.

NOTA DE COMPATIBILIDAD: pysentimiento usa internamente la clase
Trainer/TrainingArguments de transformers para hacer inferencia por
lotes. Si tu versión de transformers es muy reciente, esa API interna
puede haber cambiado y el import podría fallar; si eso ocurre, la
solución más simple es instalar pysentimiento en un entorno separado
con una versión de transformers algo más antigua (4.x), o revisar el
mensaje de error para fijar la versión que pide.
------------------------------------------------------------------------

Requisitos (instalar una sola vez en tu entorno local):
    pip install pysentimiento torch pandas tqdm

Input : dataset_unido.csv (generado por 01_preparar_datos.py)
Output: analisis_emocion_odio_ironia.csv
"""

import os
import time

import pandas as pd
from tqdm import tqdm
from pysentimiento import create_analyzer

# ------------------------------------------------------------------
# Configuración
# ------------------------------------------------------------------
INPUT_PATH = os.environ.get("INPUT_PATH", "dataset_unido.csv")
OUTPUT_PATH = os.environ.get("OUTPUT_PATH", "analisis_emocion_odio_ironia.csv")

TASKS = ["emotion", "hate_speech", "irony"]
LANG = "es"

BATCH_SIZE = int(os.environ.get("BATCH_SIZE", 32))
CHECKPOINT_EVERY = int(os.environ.get("CHECKPOINT_EVERY", 200))  # filas


def _sanitize(nombre):
    """Convierte un label de pysentimiento (p.ej. 'not ironic') en un
    nombre de columna seguro (p.ej. 'not_ironic')."""
    return nombre.strip().lower().replace(" ", "_")


def clasificar_lote(analyzers, textos):
    """Corre las 3 tareas sobre un lote de textos y devuelve una lista de
    dicts {columna: valor} -uno por texto- con las probabilidades de cada
    tarea (prob_<tarea>_<clase>) más la etiqueta que resulta de esa tarea:
    - tareas single-label  -> columna '<tarea>_etiqueta' (un solo valor)
    - tareas multi-label   -> columna '<tarea>_etiquetas' (lista separada
      por coma, o 'ninguna' si ninguna categoría superó el umbral)."""
    filas = [dict() for _ in textos]

    for task, analyzer in analyzers.items():
        salidas = analyzer.predict(textos)
        if not isinstance(salidas, list):  # por si acaso llega un solo output
            salidas = [salidas]

        for fila, salida in zip(filas, salidas):
            for label, prob in salida.probas.items():
                fila[f"prob_{task}_{_sanitize(label)}"] = prob
            if salida.is_multilabel:
                fila[f"{task}_etiquetas"] = ", ".join(salida.output) if salida.output else "ninguna"
            else:
                fila[f"{task}_etiqueta"] = salida.output

    return filas


def main():
    df = pd.read_csv(INPUT_PATH)
    n_total = len(df)
    print(f"Publicaciones a clasificar: {n_total}")
    print(f"Tareas: {', '.join(TASKS)} (lang={LANG})")

    textos = df["descripcion"].astype(str).tolist()

    # --- Retomar desde checkpoint si existe (por si una corrida anterior se cortó) ---
    checkpoint_path = OUTPUT_PATH + ".checkpoint.csv"
    filas_resultado = []
    inicio = 0

    if os.path.exists(checkpoint_path):
        parcial_previo = pd.read_csv(checkpoint_path)
        columnas_nuevas = [c for c in parcial_previo.columns if c not in df.columns]
        if len(parcial_previo) <= n_total and columnas_nuevas:
            inicio = len(parcial_previo)
            filas_resultado = parcial_previo[columnas_nuevas].to_dict(orient="records")
            print(f"Checkpoint encontrado: retomando desde la publicación {inicio}/{n_total}.")
        else:
            print("Aviso: el checkpoint no coincide con el dataset actual; se ignora y se parte de cero.")

    if inicio >= n_total:
        print("El checkpoint ya tiene todas las publicaciones clasificadas; solo falta consolidar el resultado.")

    t0 = time.time()
    analyzers = {task: create_analyzer(task=task, lang=LANG, batch_size=BATCH_SIZE) for task in TASKS}
    print(f"Analizadores cargados en {time.time() - t0:.1f}s")

    BATCHES_POR_CHECKPOINT = max(1, CHECKPOINT_EVERY // BATCH_SIZE)
    contador_batches = 0
    t0 = time.time()
    for start in tqdm(range(inicio, len(textos), BATCH_SIZE), desc="Clasificando"):
        lote = textos[start:start + BATCH_SIZE]
        filas_resultado.extend(clasificar_lote(analyzers, lote))
        contador_batches += 1

        # checkpoint incremental por si el proceso se corta
        if contador_batches % BATCHES_POR_CHECKPOINT == 0:
            parcial = df.iloc[:len(filas_resultado)].copy()
            nuevas = pd.DataFrame(filas_resultado)
            for col in nuevas.columns:
                parcial[col] = nuevas[col].values
            parcial.to_csv(checkpoint_path, index=False, encoding="utf-8-sig")

    print(f"Clasificación completa en {(time.time() - t0)/60:.1f} min")

    nuevas = pd.DataFrame(filas_resultado)
    for col in nuevas.columns:
        df[col] = nuevas[col].values

    df.to_csv(OUTPUT_PATH, index=False, encoding="utf-8-sig")
    print(f"Guardado: {OUTPUT_PATH}")

    # Limpiar checkpoint si todo salió bien
    ckpt = OUTPUT_PATH + ".checkpoint.csv"
    if os.path.exists(ckpt):
        os.remove(ckpt)

    # --- Resúmenes rápidos ---
    print("\nDistribución de emotion_etiqueta:")
    print(df["emotion_etiqueta"].value_counts())

    print("\nDistribución de irony_etiqueta:")
    print(df["irony_etiqueta"].value_counts())

    print("\nFrecuencia de cada categoría de hate_speech (multi-etiqueta, no excluyentes):")
    for label in ["hateful", "targeted", "aggressive"]:
        col = f"prob_hate_speech_{label}"
        if col in df.columns:
            n = int((df[col] > 0.5).sum())
            print(f"  {label}: {n} ({n/len(df):.1%})")


if __name__ == "__main__":
    main()
